### ISSUE TITLE
Asymmetric cryptography user authentication

### ISSUE THREAD

**Issue by bakape** [bakape/meguca#1151]
- [ ] Instead of keeping around many kinds of session like login session and user sessions use one UUID and ECDSA key pair for authentication
- [ ] When a client connects to the server for the first time, create and store private key + public key + UUID for it.
- [ ] The client can generate the cert itself, thereby protecting public key storage on the server with a captcha. The storage only needs to be done lazily after captcha authentication on first guarded resource usage.  
- [ ] Client passes cert UUID and signs all new connection attempts with it.
- [ ] If we move all traffic to websockets, session management (including antispam) becomes easier, as the client only needs to sign the connection attempt request. 
- [ ] Can replace login sessions and passwords
- [ ] Can replace captcha sessions
- [ ] Can replace post passwords
- [ ] Can be used for OP moderation rights
- [ ] Clean up certs, that haven't been used in a while
- [ ] Add cert import and export buttons to Identity settings modal

@Chiiruno Any input?

**Comment by Chiiruno** [comment]
It looks great, so far.
I'll try to think of some edge cases tomorrow.
Only immediate issue I can think of is local storage and whether that's going to reliably keep a private/public key, but if the OP is enough of a bwaka to not save the key and for some reason kill the entire browser cache, then oh well.

**Comment by Chiiruno** [comment]
Also wouldn't this kill the need for TLS and maybe even gzip for a rather large part, other than statically served HTML, etc?

**Comment by Chiiruno** [comment]
No, TLS would still be needed, not all websocket connections are guaranteed to be guarded by async crypto.

**Comment by bakape** [comment]
Only the initial connection request will be signed.

**Comment by Chiiruno** [comment]
Is there any way we can reuse TLS in this case?

**Comment by Chiiruno** [comment]
As in, instead of a custom async system, reuse TLS.

**Comment by bakape** [comment]
In a browser? Think not, but you can research that.

**Comment by Chiiruno** [comment]
Can't reuse TLS, since there's a timeout on TLS resumption.
Unless, assuming we can somehow hook into the private/public key in-browser, we use this as a temporary session.

**Comment by Chiiruno** [comment]
Would be pointless for it to be temporary though, it would nullify the point of this.
Rolling our own key pair seems to be for the best.

**Comment by Chiiruno** [comment]
Assuming threads never expire, key pair should never expire.
Speaking of things never expiring, how are we going to handle IP retention?
Just continuously delete posts in-thread other than OP?

**Comment by Chiiruno** [comment]
In that case, OP post should not have IP attached to it.

**Comment by the8472** [comment]
Some concerns about using asymmetric crypto: 

* privacy, the keys could act as user-identifiers that are more long-lived than the IP
* DoS (asymmetric operations are comparatively expensive)
* easy to get wrong when you work with the low-level primitives

**Comment by bakape** [comment]
We can provide a button to explicitly regenerate the cert and to swap it
for a different one, if they want . Does this clear the privacy concerns?

I'll need to investigate the performance penalties, but it should not be
that big of an issue, considering cert checks can be done lazily after a
valid captcha is solved. Once you do have a captcha, you can spam the cert
check on connection. Perhaps that can increment the spam score.

**Comment by bakape** [comment]
Or we could just use 1 KB of random data as the global password for anything a user does, which has to be submitted on establishing a WS connection and performing image uploads. The whole signing bit seems pointless complexity on second thought.

**Comment by the8472** [comment]
why 1kB? 256bits chosen at random gives you 128bits of collision space (birthday paradox), i.e. 2^128 possible users before collisions become likely.

**Comment by bakape** [comment]
The question is not collision space but enough entropy to use it as a
non-expiring authentication token. But yes, 1 KB is too much. 64 bytes will
suffice.
